Linux desktop fails to run with fresh install

## Steps to Reproduce

1. Install Flutter, switch the the master branch, and enable Linux support
2. `sudo apt install build-essential clang ninja-build`
3. `flutter create testapp` and `flutter run -d linux -v` inside the new project

This issue was encountered on a Dell XPS 9300 dev edition running Ubuntu 18.04

## Logs

```
~/flutter/testapp$ flutter run -d linux -v
[   +6 ms] executing: [/home/jack/flutter/flutter/] git -c log.showSignature=false log -n 1
--pretty=format:%H
[  +18 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] aa49821613372234368d2371d76d48f6250f4ee0
[        ] executing: [/home/jack/flutter/flutter/] git tag --contains HEAD
[ +121 ms] Exit code 0 from: git tag --contains HEAD
[   +1 ms] executing: [/home/jack/flutter/flutter/] git describe --match *.*.*-*.*.pre
--first-parent --long --tags
[  +99 ms] Exit code 0 from: git describe --match *.*.*-*.*.pre --first-parent --long --tags
[        ] 1.19.0-1.0.pre-113-gaa4982161
[   +4 ms] executing: [/home/jack/flutter/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/master
[        ] executing: [/home/jack/flutter/flutter/] git ls-remote --get-url origin
[   +3 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +26 ms] Unable to locate an Android SDK.
[   +6 ms] executing: [/home/jack/flutter/flutter/] git rev-parse --abbrev-ref HEAD
[   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] master
[  +23 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +15 ms] Unable to locate an Android SDK.
[   +9 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[ +110 ms] Generating
/home/jack/flutter/testapp/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistr
ant.java
[  +58 ms] Starting incremental build...
[  +11 ms] Initializing file store
[  +13 ms] Skipping target: gen_localizations
[   +7 ms] complete
[   +5 ms] Launching lib/main.dart on Linux in debug mode...
[   +5 ms] /home/jack/flutter/flutter/bin/cache/dart-sdk/bin/dart
/home/jack/flutter/flutter/bin/cache/artifacts/engine/linux-x64/frontend_server.dart.snapshot
--sdk-root /home/jack/flutter/flutter/bin/cache/artifacts/engine/common/flutter_patched_sdk/
--incremental --target=flutter --debugger-module-names
-Ddart.developer.causal_async_stacks=true --output-dill
/tmp/flutter_tools.MULLCB/flutter_tool.QWDPDN/app.dill --packages .packages
-Ddart.vm.profile=false -Ddart.vm.product=false
--bytecode-options=source-positions,local-var-info,debugger-stops,instance-field-initializers,
keep-unreachable-code,avoid-closure-call-instructions --enable-asserts --track-widget-creation
--filesystem-scheme org-dartlang-root --initialize-from-dill build/cache.dill.track.dill
[  +36 ms] Building Linux application...
[  +12 ms] <- compile package:testapp/main.dart
[   +2 ms] executing: [build/linux/debug/] cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug
/home/jack/flutter/testapp/linux
[  +18 ms] -- Configuring done
[        ] -- Generating done
[        ] -- Build files have been written to: /home/jack/flutter/testapp/build/linux/debug
[   +5 ms] executing: ninja -C build/linux/debug install
[   +5 ms] ninja: Entering directory `build/linux/debug'
[+5042 ms] [1/13] Generating
/home/jack/flutter/testapp/linux/flutter/ephemeral/libflutter_linux_glfw.so, flutter_export.h,
flutter_glfw.h, flutter_messenger.h, flutter_plugin_registrar.h,
/home/jack/flutter/testapp/linux/flutter/ephemeral/cpp_client_wrapper_glfw/engine_method_resul
t.cc,
/home/jack/flutter/testapp/linux/flutter/ephemeral/cpp_client_wrapper_glfw/standard_codec.cc,
/home/jack/flutter/testapp/linux/flutter/ephemeral/cpp_client_wrapper_glfw/plugin_registrar.cc
,
/home/jack/flutter/testapp/linux/flutter/ephemeral/cpp_client_wrapper_glfw/flutter_window_cont
roller.cc, _phony_
[   +1 ms] [   +7 ms] executing: [/home/jack/flutter/flutter/] git -c log.showSignature=false
log -n 1 --pretty=format:%H
[        ] [  +17 ms] Exit code 0 from: git -c log.showSignature=false log -n 1
--pretty=format:%H
[        ] [        ] aa49821613372234368d2371d76d48f6250f4ee0
[        ] [        ] executing: [/home/jack/flutter/flutter/] git tag --contains HEAD
[        ] [ +136 ms] Exit code 0 from: git tag --contains HEAD
[        ] [   +1 ms] executing: [/home/jack/flutter/flutter/] git describe --match
*.*.*-*.*.pre --first-parent --long --tags
[        ] [ +126 ms] Exit code 0 from: git describe --match *.*.*-*.*.pre --first-parent
--long --tags
[        ] [        ] 1.19.0-1.0.pre-113-gaa4982161
[        ] [   +9 ms] executing: [/home/jack/flutter/flutter/] git rev-parse --abbrev-ref
--symbolic @{u}
[        ] [   +5 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] [        ] origin/master
[        ] [        ] executing: [/home/jack/flutter/flutter/] git ls-remote --get-url origin
[        ] [   +5 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] [        ] https://github.com/flutter/flutter.git
[        ] [  +60 ms] Unable to locate an Android SDK.
[        ] [  +10 ms] executing: [/home/jack/flutter/flutter/] git rev-parse --abbrev-ref HEAD
[        ] [   +5 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] [        ] master
[        ] [  +35 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[        ] [   +3 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required,
skipping update.
[        ] [   +6 ms] Artifact Instance of 'MaterialFonts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'GradleWrapper' is not required, skipping update.
[        ] [        ] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FlutterSdk' is not required, skipping update.
[        ] [        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping
update.
[        ] [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required,
skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FontSubsetArtifacts' is not required, skipping
update.
[        ] [  +84 ms] Initializing file store
[        ] [  +11 ms] Done initializing file store
[        ] [  +30 ms] Skipping target: gen_localizations
[        ] [+1504 ms] Skipping target: kernel_snapshot
[        ] [+2247 ms] Skipping target: unpack_linux
[        ] [ +236 ms] Skipping target: debug_bundle_linux_assets
[        ] [        ] Persisting file store
[        ] [   +6 ms] Done persisting file store
[        ] [   +4 ms] build succeeded.
[        ] [   +4 ms] "flutter assemble" took 4,154ms.
[ +112 ms] [2/5] Building CXX object
CMakeFiles/testapp.dir/flutter/generated_plugin_registrant.cc.o
[ +202 ms] [3/5] Building CXX object CMakeFiles/testapp.dir/main.cc.o
[  +53 ms] [4/5] Linking CXX executable testapp
[        ] FAILED: testapp 
[        ] : && /usr/bin/clang++  -g   CMakeFiles/testapp.dir/main.cc.o
CMakeFiles/testapp.dir/window_configuration.cc.o
CMakeFiles/testapp.dir/flutter/generated_plugin_registrant.cc.o  -o testapp
-Wl,-rpath,/home/jack/flutter/testapp/linux/flutter/ephemeral:
flutter/libflutter_wrapper_app.a
/home/jack/flutter/testapp/linux/flutter/ephemeral/libflutter_linux_glfw.so && :
[        ] CMakeFiles/testapp.dir/main.cc.o: In function `(anonymous
namespace)::RunHeadless(std::__cxx11::basic_string<char, std::char_traits<char>,
std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>,
std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char,
std::char_traits<char>, std::allocator<char> >,
std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >
> > const&)':
[        ] /home/jack/flutter/testapp/linux/main.cc:20: undefined reference to
`flutter::FlutterEngine::FlutterEngine()'
[        ] /home/jack/flutter/testapp/linux/main.cc:21: undefined reference to
`flutter::FlutterEngine::Start(std::__cxx11::basic_string<char, std::char_traits<char>,
std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>,
std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char,
std::char_traits<char>, std::allocator<char> >,
std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >
> > const&)'
[        ] /home/jack/flutter/testapp/linux/main.cc:24: undefined reference to
`flutter::FlutterEngine::RunEventLoopWithTimeout(std::chrono::duration<long, std::ratio<1l,
1000l> >)'
[        ] /home/jack/flutter/testapp/linux/main.cc:26: undefined reference to
`flutter::FlutterEngine::~FlutterEngine()'
[        ] clang: error: linker command failed with exit code 1 (use -v to see invocation)
[        ] ninja: build stopped: subcommand failed.
[   +6 ms] Building Linux application... (completed in 5.5s)
[        ] Exception: Build process failed
[   +1 ms] "flutter run" took 5,765ms.

#0      throwToolExit (package:flutter_tools/src/base/common.dart:14:3)
#1      RunCommand.runCommand (package:flutter_tools/src/commands/run.dart:569:7)
<asynchronous suspension>
#2      FlutterCommand.verifyThenRunCommand
(package:flutter_tools/src/runner/flutter_command.dart:797:18)
<asynchronous suspension>
#3      FlutterCommand.run.<anonymous closure>
(package:flutter_tools/src/runner/flutter_command.dart:694:33)
<asynchronous suspension>
#4      FlutterCommand.run.<anonymous closure>
(package:flutter_tools/src/runner/flutter_command.dart)
#5      AppContext.run.<anonymous closure>
(package:flutter_tools/src/base/context.dart:150:29)
#6      _rootRun (dart:async/zone.dart:1190:13)
#7      _CustomZone.run (dart:async/zone.dart:1093:19)
#8      _runZoned (dart:async/zone.dart:1630:10)
#9      runZoned (dart:async/zone.dart:1550:10)
#10     AppContext.run (package:flutter_tools/src/base/context.dart:149:18)
#11     FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:684:20)
#12     CommandRunner.runCommand (package:args/command_runner.dart:197:27)
#13     FlutterCommandRunner.runCommand.<anonymous closure>
(package:flutter_tools/src/runner/flutter_command_runner.dart:339:21)
#14     _rootRunUnary (dart:async/zone.dart:1198:47)
#15     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#16     _FutureListener.handleValue (dart:async/future_impl.dart:143:18)
#17     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:696:45)
#18     Future._propagateToListeners (dart:async/future_impl.dart:725:32)
#19     Future._completeWithValue (dart:async/future_impl.dart:529:5)
#20     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:40:15)
#21     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:311:13)
#22     FlutterVersion.checkFlutterVersionFreshness (package:flutter_tools/src/version.dart)
#23     _rootRunUnary (dart:async/zone.dart:1198:47)
#24     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#25     _FutureListener.handleValue (dart:async/future_impl.dart:143:18)
#26     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:696:45)
#27     Future._propagateToListeners (dart:async/future_impl.dart:725:32)
#28     Future._completeWithValue (dart:async/future_impl.dart:529:5)
#29     Future._asyncCompleteWithValue.<anonymous closure> (dart:async/future_impl.dart:567:7)
#30     _rootRun (dart:async/zone.dart:1190:13)
#31     _CustomZone.run (dart:async/zone.dart:1093:19)
#32     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
#33     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
#34     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#35     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#36     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#37     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:169:5)



[   +4 ms] Running shutdown hooks
[        ] Shutdown hook priority 4
[   +2 ms] Shutdown hooks complete
```

## Flutter doctor

```
[✓] Flutter (Channel master, 1.19.0-2.0.pre.113, on Linux, locale en_US.UTF-8)
    • Flutter version 1.19.0-2.0.pre.113 at /home/jack/flutter/flutter
    • Framework revision aa49821613 (53 minutes ago), 2020-05-18 17:32:02 -0400
    • Engine revision 4b1a70e6a2
    • Dart version 2.9.0 (build 2.9.0-9.0.dev 40f7a11d89)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed
      instructions).
      If the Android SDK has been installed to a custom location, set ANDROID_SDK_ROOT to that
      location.
      You may also want to add it to your PATH environment variable.


[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed
      instructions).

[✓] VS Code (version 1.45.1)
    • VS Code at /usr/share/code
    • Flutter extension version 3.10.2

[✓] Connected device (1 available)
    • Linux • Linux • linux-x64 • Linux

! Doctor found issues in 2 categories.
```

--- a/packages/flutter_tools/templates/app/linux.tmpl/flutter/CMakeLists.txt
+++ b/packages/flutter_tools/templates/app/linux.tmpl/flutter/CMakeLists.txt
@@ -52,6 +52,7 @@ list(APPEND CPP_WRAPPER_SOURCES_PLUGIN
 )
 list_prepend(CPP_WRAPPER_SOURCES_PLUGIN "${WRAPPER_ROOT}/")
 list(APPEND CPP_WRAPPER_SOURCES_APP
+  "flutter_engine.cc"
   "flutter_window_controller.cc"
 )
 list_prepend(CPP_WRAPPER_SOURCES_APP "${WRAPPER_ROOT}/")

